### DISCUSSION TITLE
Compositions vs Helm charts

### DISCUSSION THREAD

**Discussion by dudicoco** [crossplane/crossplane#3383] [Q&A]
Hi,

I've spent a while trying to wrap my head around the compositions docs after reading them multiple times, the concept is just not very user friendly.

Since for me (and possibly others) creating a helm chart is much easier than dealing with compositions, I'm wondering if there is any advantage in using compositions over grouping your resources in a helm chart?

Would a composition create the managed resources according to a dependency order? 

Is the patching functionality provide anything that can't be done with standard managed resources?

Thanks

**Comment by bobh66** [comment]
> Since for me (and possibly others) creating a helm chart is much easier than dealing with compositions, I'm wondering if there is any advantage in using compositions over grouping your resources in a helm chart?

There are a number of advantages to using compositions over Helm charts but the two most obvious to me are:
- Compositions are actively reconciled to maintain state. Helm charts are are a one-shot implementation method - external changes to the deployed resources are not detected or corrected automatically
- Compositions allow retrieving data from one deployed resource and using that data in other resources - so you can create a VPC and retrieve the (generated) VPC ID for use in creating Subnets, etc.  All helm chart data has to be available at deployment time, it cannot be retrieved and reused during deployment

> Would a composition create the managed resources according to a dependency order?

By default Crossplane and Compositions in particular, are eventually consistent.  The deployment order can be influenced using required fields and policy.fromFieldPath: Required

> Is the patching functionality provide anything that can't be done with standard managed resources?

Patching is only supported in Compositions - standard managed resources are just declarative Kubernetes CRDs

**Reply by dudicoco** [reply]
Thanks for the info @bobh66.

> Compositions are actively reconciled to maintain state. Helm charts are are a one-shot implementation method - external changes to the deployed resources are not detected or corrected automatically

I was under the impression that all managed resources are constantly being reconciled, even if they are not defined within a composition. Are you saying that this is not the case? If so, is it documented anywhere? 

> Compositions allow retrieving data from one deployed resource and using that data in other resources - so you can create a VPC and retrieve the (generated) VPC ID for use in creating Subnets, etc. All helm chart data has to be available at deployment time, it cannot be retrieved and reused during deployment

How is that different than the `vpcIdRef` argument which is available within the managed resources? https://github.com/crossplane-contrib/provider-aws/blob/5a380ad73a80fc1dad5e91bc7d4d5aa0f14e6081/examples/ec2/subnet.yaml#L11

**Comment by bobh66** [comment]
> I was under the impression that all managed resources are constantly being reconciled, even if they are not defined within a composition

That's true - if you deploy one or more Managed Resources via a Helm chart those individual resources will be reconciled by the Crossplane provider that is responsible for them.  However any non-Managed Resources will not be reconciled.

> How is that different than the vpcIdRef argument which is available within the managed resources? 

The *IdRef parameters in a Managed Resource can be determined by the provider but that's a very small amount of data compared to what Composition patching can do.  The *IdRef parameter needs to be defined in the MR for the provider to process it, whereas patching can generate a value for any field in the spec using data that is external to the MR.  For example the oidcIssuer of an EKS Cluster can be patched into a Role that needs it for the assumeRoleDocumentPolicy.   There is no way to do that in Helm.

You certainly can use Helm to orchestrate Managed Resources and there may be scenarios where that is sufficient to meet the requirements.  I think it limits the ability to generate resource abstractions that encapsulate all of the associated complexity compared to what you can achieve with Compositions, and it doesn't necessarily provide for re-usable resource abstractions, but that may not be your goal and is not a reason not to do it in a specific situation.

**Reply by dudicoco** [reply]
> The *IdRef parameters in a Managed Resource can be determined by the provider but that's a very small amount of data compared to what Composition patching can do. The *IdRef parameter needs to be defined in the MR for the provider to process it, whereas patching can generate a value for any field in the spec using data that is external to the MR. For example the oidcIssuer of an EKS Cluster can be patched into a Role that needs it for the assumeRoleDocumentPolicy. There is no way to do that in Helm.

Is https://github.com/crossplane/crossplane/pull/3008 going to make this possible with managed resources outside of composition as well?

**Comment by bobh66** [comment]
> Is https://github.com/crossplane/crossplane/pull/3008 going to make this possible with managed resources outside of composition as well?

No - that refers to patching within a Composition

**Comment by jbw976** [comment]
Good question @dudicoco, I've heard a couple other folks bringing this up recently too.  Thanks @bobh66 for providing some good answers already too!

A key distinction I look at here is that compositions essentially enables you to create your own infrastructure API - and an API is super important here and very beneficial IMHO.

For instance, your devs can create instances of AcmeDatabase or AcmeBucket - that is the typed API they interact with, instead of generic (untyped) Helm charts.

When you normalize on an API, you get a lot of beneficial behaviors.  For example, any tool can interact with the API now, so you have a lot of portability.  You can hit it with kubectl or any of the k8s tools ecosystems, you can make a HTTP call, you can call it from programming languages like golang or python, etc. - there's a lot of flexibility that is enabled by creating an API for your infrastructure. Moving the implementation details to server side, as opposed to client side operations, also increases this portability and flexibility.

I also like the permissions model when using an API - you can give your devs permissions to interact only with AcmeDatabase.  That's it - not the underlying managed resources, not the cloud provider managed services, just AcmeDatabase. The permission model is raised up to the level of abstraction/API, which is quite nice.

There's other benefits that come from compositions in Crossplane (@bobh66 mentioned a few), but to me a lot of the benefits come from the general approach of enabling you to abstract your infrastructure and treat it as an API 💪

**Reply by dudicoco** [reply]
Thanks for the info @jbw976.

Regarding the benefits of creating infrastructure APIs vs Helm charts:
1. You can have a server side helm controller such as the crossplane controller, flux controller and others which somewhat moves the implementation to server side
2. You can validate Helm charts values structure with a json schema: https://helm.sh/docs/topics/charts/#schema-files
3. Why would you want devs to directly create infra rather than going through GIT and letting your CI create the infra?
4.  I think it might be possible to restrict permissions on who can deploy specific charts with the server side helm controllers

My main concern with compositions is that they are much more complicated to work with than helm charts and are much less powerful than go templating (if else statements, for loops [sprig functions](http://masterminds.github.io/sprig/) etc.)
